### PR TITLE
Feature/update existing process document endpoint

### PR DESCRIPTION
**Pull request by MirelaGeorgiou**

### CODE REVIEW

File: mat-process-api.Tests/V1/Controllers/ProcessDataControllerAcceptanceTests.cs [hunk 1]
```diff
@@ -73,6 +73,36 @@ public void controller_end_to_end_test()
             Assert.IsInstanceOf<MatProcessData>(getProcessDataResponse.ProcessData);
             Assert.AreEqual(processRef,getProcessDataResponse.ProcessData.Id);
         }
+        [TestCase("00000000-0000-0000-0000-000000000000")]
+        [TestCase("00000000-dd23-0000-abcd-000000000000")]
+        [TestCase("2539ca27-12c0-e811-a96c-002248072cb4")]
+        public void update_process_controller_end_to_end_test(string processRef)
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/Controllers/ProcessDataControllerAcceptanceTests.cs:79]: Nice to have: Remove the 'TestCase'es and replace with: string processRef = _faker.Random.Guid().ToString(); in order to maintain consistency across tests. The triangulation will be ensured by the same principle that makes a Guid good for being a unique identifier.

Comment (Duslerke) [inline mat-process-api.Tests/V1/Controllers/ProcessDataControllerAcceptanceTests.cs:79]: I feel that we should include at least two more tests, to test for failure. We'd need an acceptance test for:
- when the input is invalid, and then check if we get correct 400 response
- we also should check if we get the correct response when we try to update the non-existing document. so that would be a 404 (resource not found).
- this one is optional. we could check whether we get correctly formatted 500 response upon any other error, but it's sort of default behaviour that this response will be returned... unless we would want to check whether our custom message has been appended to the error itself.

File: mat-process-api.Tests/V1/Controllers/ProcessDataControllerAcceptanceTests.cs [hunk 2]
```diff
@@ -73,6 +73,36 @@ public void controller_end_to_end_test()
...
+            _dbcontext.getCollection().InsertOne(BsonDocument.Parse(JsonConvert.SerializeObject(dataToInsert)));
+            //fields to update
+            var dataToUpdate = new MatUpdateProcessData();
+            dataToUpdate.Id = dataToInsert.Id;
+            dataToUpdate.DateLastModified = DateTime.Now;
+            dataToUpdate.PreProcessData = new
+            {
+                randomField = "abc"
+            };
+            var request = new UpdateProcessDataRequest() { processDataToUpdate = dataToUpdate };
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/Controllers/ProcessDataControllerAcceptanceTests.cs]: Nice to have: 
the method in the 'MatProcessDataHelper' class that would generate requests for us - would make reading tests much easier as well as it keep everything more organized and consistent across api.

File: mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs [hunk 3]
```diff
@@ -74,21 +79,157 @@ public void given_a_processRef_when_getprocessdata_method_is_called_it_then_call
...
+            //act
+            var actualResponse = _processDataController.UpdateExistingProcessDocument(request);
+            var result = (ObjectResult)actualResponse;
+            var resultContent = result.Value;
+            //assert
+            Assert.NotNull(actualResponse);
+            Assert.NotNull(result);
+            Assert.NotNull(resultContent);
+            Assert.AreEqual(errorResponse, resultContent);
+            Assert.AreEqual(200, result.StatusCode);
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs:202]: Maybe we could communicate to the client that they made a mistake in providing non-existent Guid by saying 404 'document not found'?

My reasoning is - when they make a request to Update the document, if document does not exist, then their request fails to be fulfilled. Hence I think it's more appropriate to return an error response rather than 200 success one. The 4xx is for client errors, so it's fitting.

File: mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs [hunk 4]
```diff
@@ -74,21 +79,157 @@ public void given_a_processRef_when_getprocessdata_method_is_called_it_then_call
...
+
+            _mockValidatorUpdate.Setup(x => x.Validate(request)).Returns(fakeValidationResult);
+            var response = _processDataController.UpdateExistingProcessDocument(request);
+            var okResult = (ObjectResult)response;
+            var resultContent = okResult.Value;
+
+            Assert.NotNull(response);
+            Assert.NotNull(okResult);
+            Assert.NotNull(resultContent);
+            Assert.AreEqual(JsonConvert.SerializeObject(response), JsonConvert.SerializeObject(okResult));
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs]: I'm pretty sure that after passing the `Assert.NotNull(okResult);` assertion.

This assertion: 
`Assert.AreEqual(JsonConvert.SerializeObject(response), JsonConvert.SerializeObject(okResult));` can't fail. Because you're comparing the exact same object with itself, except the other copy is typecasted. If the typecast fails, I'm pretty sure you'd see it here: `Assert.NotNull(okResult);`, as this assertion would be the first one to fail, and stop the test.

I'm not sure if the need that check, since it wouldn't be able to show anything due to it being unable to fail.

File: mat-process-api/V1/Factories/ProcessDataFactory.cs [hunk 5]
```diff
@@ -24,9 +26,56 @@ public static MatProcessData CreateProcessDataObject(BsonDocument bsonResult)
             }
         }
 
+        public static UpdateDefinition<BsonDocument> PrepareFieldsToBeUpdated(MatUpdateProcessData matProcessData)
```
Comment (Duslerke) [inline mat-process-api/V1/Factories/ProcessDataFactory.cs]: It would be good to move this functionality into separate class like 'QueryBuilder/GatewayHelper/...'.

It doesn't feel like this belongs here. I think that Factory's job is to convert or map one object to another (like request to domain). It also maps from Bson back to domain (but it's the same idea).

While this functionality is more about building up a list of instructions for mongo driver so that document in db could be updated. Very different responsibilities, don't think they fit in with single responsibility principle. I could be wrong though.

File: mat-process-api.Tests/V1/Factories/ProcessDataFactoryTest.cs [hunk 6]
```diff
@@ -93,5 +101,66 @@ public void given_postInitialProcessDocumentRequest_object_when_createProcessDat
             Assert.AreEqual(0, domainObject.ProcessData.GetType().GetProperties().Count());
             Assert.AreEqual(0, domainObject.PostProcessData.GetType().GetProperties().Count());
         }
+
+        #region Update process data
+        [Test]
+        public void can_check_for_fields_to_be_updated_from_a_request_object_and_prepare_update_definition()
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/Factories/ProcessDataFactoryTest.cs]: Would be nice if tests were moved into a corresponding class as well like 'GatewayHelperTests' or something similar. I feel like the functionalities mixed within the factory files don't match well, when it comes to single responsibility.

File: mat-process-api.Tests/V1/Factories/ProcessDataFactoryTest.cs [hunk 7]
```diff
@@ -93,5 +101,66 @@ public void given_postInitialProcessDocumentRequest_object_when_createProcessDat
...
+            var documentSerializer = serializerRegistry.GetSerializer<BsonDocument>();
+            //compare what the update definitions render to
+            Assert.AreEqual(expectedUpdateDefinition.Render(documentSerializer, serializerRegistry).ToString(),
+                processDataUpdateDefinition.Render(documentSerializer,serializerRegistry).ToString());
+        }
+
+        [TestCase("test")]
+        [TestCase("test-123--")]
+        [TestCase("4455")]
+        public void can_check_for_fields_to_be_updated_from_a_request_object_correctly(string firstField)
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/Factories/ProcessDataFactoryTest.cs]: Need to remove these TestCases as they're not being used, _faker does the job.

File: mat-process-api.Tests/V1/Gateways/ProcessesGatewayTests.cs [hunk 8]
```diff
@@ -91,6 +92,54 @@ public void test_that_gateway_return_mat_process_object_if_no_match_is_found()
             Assert.IsInstanceOf<MatProcessData>(result);
         }
 
+        #region Update Process
+        [Test]
+        public void test_that_object_can_be_successfully_updated()
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/Gateways/ProcessesGatewayTests.cs:97]: This test doesn't test whether the object was successfully updated or not. It only tests whether the gateway returns the response we expect. (Which as a test we would need a anyway, so it's best to rename it.)

In order to test whether the gateway has updated the object, you'd need to run a get query after you have updated the object, and then compare if the object you retrieved from database matches your expected object.

Otherwise there's no way of knowing if the object in the db was updated, for all you know you only updated it in RAM memory and returned it back.

Comment (Duslerke) [reply mat-process-api.Tests/V1/Gateways/ProcessesGatewayTests.cs:97]: So it's 2 tests we need: 
1) to check if the object was indeed updated within the database.
2) to see whether the gateway returns that object correctly.

File: mat-process-api.Tests/V1/UseCase/ListTransactionsUsecaseTests.cs [hunk 9]
```diff
@@ -13,8 +13,8 @@ namespace UnitTests.V1.UseCase
     [TestFixture]
     public class ListTransactionsUsecaseTests
     {
-        private ListTransactionsUsecase _classUnderTest;
-        private Mock<ITransactionsGateway> _transactionsGateway;
+       // private ListTransactionsUsecase _classUnderTest;
+        //private Mock<ITransactionsGateway> _transactionsGateway;
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/UseCase/ListTransactionsUsecaseTests.cs]: It's not for this PR, but I've been thinking for a while now that we should remove all the transactions api code by now - it's getting hard to find the tests you need due to transaction api namespaces, these tests are being run for no reason and waste time. Would be good to have some refactor time at some point. Writing it here now, as I keep forgetting to bring this up.

File: mat-process-api/V1/Boundary/UpdateProcessDataRequest.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using mat_process_api.V1.Domain;
+
+namespace mat_process_api.V1.Boundary
+{
+    public class UpdateProcessDataRequest
+    {
+        public MatUpdateProcessData processDataToUpdate { get; set; }
```
Comment (Duslerke) [inline mat-process-api/V1/Boundary/UpdateProcessDataRequest.cs:13]: I'm wondering whether we could put the contents of 'MatUpdateProcessData' class into this ('UpdateProcessDataRequest') class? I'm not sure if we need 1 extra layer of wrapping.

Uhm, what I mean is, why is the update data wrapped up inside 'MatUpdateProcessData', when it could be like this?:

public class UpdateProcessDataRequest
{
   [JsonProperty("_id")]
   [BsonRepresentation(BsonType.ObjectId)]
        public string Id { get; set; }
        [JsonProperty("dateCompleted")]
   [BsonElement("dateCompleted")]
   public DateTime DateCompleted { get; set; }
   [JsonProperty("dateLastModified")]
   [BsonElement("dateLastModified")]
   public DateTime DateLastModified { get; set; } = DateTime.Now
   [JsonProperty("processDataAvailable")]
   [BsonElement("processDataAvailable")]
   public bool ProcessDataAvailable { get; set; }
   [JsonProperty("dataSchemaVersion")]
   [BsonElement("dataSchemaVersion")]
   public int ProcessDataSchemaVersion { get; set; }
        [JsonProperty("processStage")]
        [BsonElement("processStage")]
        public string ProcessStage { get; set; }
        [JsonProperty("preProcessData")]
        [BsonElement("preProcessData")]
        public object PreProcessData { get; set; }
        [JsonProperty("processData")]
        [BsonElement("processData")]
        public object ProcessData { get; set; }
        [JsonProperty("postProcessData")]
        [BsonElement("postProcessData")]
        public object PostProcessData { get; set; }
}

And then whatever properties are not in this object do not get mapped into this object by the middleware itself. 

Also I'm pretty sure this one extra wrap up might require of the front end to also wrap their result up 1 extra time...unless they already do it. Need to try to test this through swagger, if this requires an extra wrap up or not.

File: mat-process-api/V1/Boundary/UpdateProcessDataResponse.cs [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace mat_process_api.V1.Boundary
+{
+    public class UpdateProcessDataResponse
+    {
+        public UpdateProcessDataRequest Request { get; set; }
+        public MatProcessData ProcessData { get; set; }
```
Comment (Duslerke) [inline mat-process-api/V1/Boundary/UpdateProcessDataResponse.cs]: I'm wondering whether the name of this field is clear enough 'ProcessData'... as in is it clear enough that this is how the updated process data object looks like. It could be called something among the lines of 'UpdatedDocument' or similar... to make it a bit clearer.

It would be good to have second opinions on this one, maybe it's just me.

File: mat-process-api/V1/Controllers/ProcessDataController.cs [hunk 12]
```diff
@@ -54,21 +58,40 @@ public IActionResult GetProcessData(string processRef)
         /// If processData needs to be update, it's instead replaced with a full new one.
         /// Returning flat out 200 on successful update
         /// </summary>
-        /// <param name="processDataJSON"></param>
+        /// <param name="updateRequest"></param>
         /// <returns></returns>
-        [HttpPatch]
+        [HttpPut]
```
Comment (Duslerke) [inline mat-process-api/V1/Controllers/ProcessDataController.cs]: According to our Playbook standards, we should be using "HttpPatch":
https://github.com/LBHackney-IT/API-Playbook/blob/master/api-guidelines/http.md#patch

From the playbook:
"PATCH requests are used to update parts of single resources,..."
"PUT requests are used to update entire resources. ..."

This because we're updating some fields of the document, rather than replacing the entire document.

File: mat-process-api/V1/Controllers/ProcessDataController.cs [hunk 13]
```diff
@@ -54,21 +58,40 @@ public IActionResult GetProcessData(string processRef)
...
+
+                try
+                {
+                    _logger.LogInformation($"Update ProcessData request for process ID {updateRequest.processDataToUpdate.Id}");
+                    var result = _processDataUsecase.ExecuteUpdate(updateRequest);
+                    return Ok(result);
+                }
+                catch (DocumentNotFound ex)
+                {
+                    return StatusCode(200, $"Document with reference {updateRequest.processDataToUpdate.Id} was not found in the database." +
```
Comment (Duslerke) [inline mat-process-api/V1/Controllers/ProcessDataController.cs]: I think it should be this, stated reasoning next to the tests: 404 | Not found - the resource is not found.
Also, according to the playbook:
"PATCH requests are usually not robust against non-existence of resource instances", which implies that we should use an error response rather than success one.

File: mat-process-api/V1/Controllers/ProcessDataController.cs [hunk 14]
```diff
@@ -54,21 +58,40 @@ public IActionResult GetProcessData(string processRef)
...
+                }
+            }
+            return BadRequest(isRequestValid.Errors);
         }
 
         /// <summary>
         /// Creates an intial process JSON document in the database.
         /// Upon creating a resource returns 200 Ok
         /// </summary>
-        /// <param name="processObject"></param>
```
Comment (Duslerke) [inline mat-process-api/V1/Controllers/ProcessDataController.cs:71]: I think we need to provide some sort of parameter name for the sake of swagger documentation.

File: mat-process-api/V1/Gateways/ProcessDataGateway.cs [hunk 15]
```diff
@@ -27,7 +28,24 @@ public MatProcessData GetProcessData(string processRef)
...
+        {
+            //return updated document
+            var options = new FindOneAndUpdateOptions<BsonDocument>();
+            options.ReturnDocument = ReturnDocument.After;
+            //query by
+            var filter = Builders<BsonDocument>.Filter.Eq("_id", id);
+            //find and update
+            var result = matDbContext.getCollection().FindOneAndUpdate(filter, updateDefinition, options);
+            //if empty result, the document to be updated was not found
+            if(result == null)
```
Comment (Duslerke) [inline mat-process-api/V1/Gateways/ProcessDataGateway.cs:45]: Not sure how this works - doesn't the attempt at update against non-existent resource make mongo driver throw an error? Or does it just return 'null' if the attempt fails?

File: mat-process-api/V1/Validators/UpdateProcessDocumentRequestValidator.cs [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+                return true;
+            }
+            else if (request.PreProcessData != null)
+            {
+                return true;
+            }
+            else if (request.ProcessDataSchemaVersion != 0)
+            {
+                return true;
+            }
```
Comment (Duslerke) [inline mat-process-api/V1/Validators/UpdateProcessDocumentRequestValidator.cs:47]: What happens if I make a request like this:
`request.PostProcessData = null`
`request.PreProcessData = null`
`request.ProcessData = null`  , but:
`request.ProcessDataSchemaVersion = 2`?
(and the rest of properties null)

This request would pass through the validation, without ever checking if the last processData child object is null or not.

Is it ok if I provide none of the 'ProcessData' objects, but only a 'SchemaVersion'?

File: mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs [hunk 17]
```diff
@@ -58,10 +61,12 @@ public void test_gateway_is_called()
             mockMatGateway.Verify(x => x.GetProcessData(processRef));
         }
 
-        public void verif_gateway_calls_database_with_parameters()
+        [TestCase("123")]
+        [TestCase("acb5")]
+        [TestCase("aa-bb-123")]
+        public void get_process_data_verify_gateway_calls_database_with_parameters(string processRef)
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs:67]: We don't need the TestCase'es, we can use faker instead.

File: docker-compose.yml [hunk 18]
```diff
@@ -6,6 +6,10 @@ services:
     build:
       context: mat-process-api/
       dockerfile: ./Dockerfile
+    environment:
+        - MONGO_CONN_STRING=mongodb://mongo-db:27017/?gssapiServiceName=mongodb
+    links:
+      - mongo-db
```
Comment (Duslerke) [inline docker-compose.yml]: Why did the docker file need updating?

**Review by Duslerke — CHANGES_REQUESTED** [review]
Didn't finish looking at the usecase tests, but this will be enough for now.
Apart from these few things, everything seemed really good.

**Comment by Duslerke** [comment]
If we're short on time aesthetics such as using faker rather than TestCases .. and similar can be ignored for sure.

File: mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs (hunk 3 above)
Comment (MirelaGeorgiou) [reply mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs:202]: In my opinion 404 is a status code you use when a technical resource is missing, not a record in a database. 404 signifies that there was an issue with the request itself (a resource is missing). In our case, the request is successful, there is just no matches in the DB.

File: mat-process-api.Tests/V1/Factories/ProcessDataFactoryTest.cs (hunk 7 above)
Comment (MirelaGeorgiou) [reply mat-process-api.Tests/V1/Factories/ProcessDataFactoryTest.cs]: It has been previously explained in Hackney by a digital agency that testing with multiple values to make the test more general is the way we do triangulation. Although faker generates a different value every time, when tests are run in CI/CD, they run once -> we test against single value and not multiple.

File: mat-process-api.Tests/V1/Gateways/ProcessesGatewayTests.cs (hunk 8 above)
Comment (MirelaGeorgiou) [reply mat-process-api.Tests/V1/Gateways/ProcessesGatewayTests.cs:97]: No you don't. The 'Find and Update' query run returns the updated object from the database.

File: mat-process-api.Tests/V1/UseCase/ListTransactionsUsecaseTests.cs (hunk 9 above)
Comment (MirelaGeorgiou) [reply mat-process-api.Tests/V1/UseCase/ListTransactionsUsecaseTests.cs]: They were left there for reference for people who might have needed to refer to examples. I am happy to remove them.

File: mat-process-api/V1/Validators/UpdateProcessDocumentRequestValidator.cs (hunk 16 above)
Comment (Duslerke) [reply mat-process-api/V1/Validators/UpdateProcessDocumentRequestValidator.cs:47]: If it's not intended behaviour, then the reason it passed the tests was because there weren't enough tests to cover enough cases.
Or you could make a seperate rule involving only the 'ProcessData', 'PreProcessData, ..., null checks - in this case you'd need less tests.

File: mat-process-api/V1/Gateways/ProcessDataGateway.cs (hunk 15 above)
Comment (MirelaGeorgiou) [reply mat-process-api/V1/Gateways/ProcessDataGateway.cs:45]: If document doesn't exist, it's null it returns

File: mat-process-api/V1/Validators/UpdateProcessDocumentRequestValidator.cs (hunk 16 above)
Comment (MirelaGeorgiou) [reply mat-process-api/V1/Validators/UpdateProcessDocumentRequestValidator.cs:47]: Yes

File: docker-compose.yml (hunk 18 above)
Comment (MirelaGeorgiou) [reply docker-compose.yml]: I committed this by mistake, will remove it.

File: mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs (hunk 4 above)
Comment (MirelaGeorgiou) [reply mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs]: fair point

File: mat-process-api.Tests/V1/Controllers/ProcessDataControllerAcceptanceTests.cs (hunk 2 above)
Comment (MirelaGeorgiou) [reply mat-process-api.Tests/V1/Controllers/ProcessDataControllerAcceptanceTests.cs]: True, but this gave the opportunity to use different update request objects across tests v(to send different fields for update), so not sure if in this case having it more neat is more important then testing different update combinations.

File: mat-process-api.Tests/V1/Factories/ProcessDataFactoryTest.cs (hunk 6 above)
Comment (MirelaGeorgiou) [reply mat-process-api.Tests/V1/Factories/ProcessDataFactoryTest.cs]: Good idea, but those are not Gateway helpers necessarily. I will move them to update process data helper class.

File: mat-process-api.Tests/V1/UseCase/ListTransactionsUsecaseTests.cs (hunk 9 above)
Comment (MirelaGeorgiou) [reply mat-process-api.Tests/V1/UseCase/ListTransactionsUsecaseTests.cs]: I removed them.

File: mat-process-api/V1/Controllers/ProcessDataController.cs (hunk 12 above)
Comment (MirelaGeorgiou) [reply mat-process-api/V1/Controllers/ProcessDataController.cs]: It's currently a PUT as they are sending the whole update process data object. This might change if we change it to individual fields being sent, rather then wrapped in an object.

File: mat-process-api/V1/Boundary/UpdateProcessDataResponse.cs (hunk 11 above)
Comment (MirelaGeorgiou) [reply mat-process-api/V1/Boundary/UpdateProcessDataResponse.cs]: Good suggestion.

File: mat-process-api/V1/Controllers/ProcessDataController.cs (hunk 12 above)
Comment (Duslerke) [reply mat-process-api/V1/Controllers/ProcessDataController.cs]: So, if we're being provided the whole object. Why are the making it complicated for ourselves and updating individual fields in the gateway? 

We can simply replace the whole document - that way we won't need any query builder, which just further complicates the program.

File: mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs (hunk 3 above)
Comment (Duslerke) [reply mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs:202]: Yeah it makes sense, I'm not sure about the 200 though. 

I feel like if you are trying make a request on non-existent resource, then you've made some sort of error and should be notified with an some error status code that you did something you shouldn't have even tried.

But it's debatable on the internet for either case.... We can leave it as is - it's not very relevant right now anyway. 

But this leaves an issue open of that our Playbook is not robust enough to enforce a standard.
In this case it just points to this:
https://tools.ietf.org/html/rfc6902#page-8
which points to this: 
https://tools.ietf.org/html/rfc5789#section-2.2
which says:

> Resource not found:  Can be specified with a 404 (Not Found) status
>       code when the client attempted to apply a patch document to a non-
>       existent resource, but the patch document chosen cannot be applied
>       to a non-existent resource.

All it says is you "Can" do it, but it means you can also not do it. Leaving you do as you please. Though I believe that we should have stuck with playbook suggestion, since we all explicitly agreed on it. I don't think that trying to do something that shouldn't be done _(calling patch on non-existent resource)_ is called a success.
Need to raise this issue on API playbook at some point.

File: mat-process-api.Tests/V1/Gateways/ProcessesGatewayTests.cs (hunk 8 above)
Comment (Duslerke) [reply mat-process-api.Tests/V1/Gateways/ProcessesGatewayTests.cs:97]: I understand that that's the implementation, but the idea of TDD that the test goes first and then you add the implementation to make it pass.
In order to make the gateway return the object, you needed to add these lines to the implementation:
`var options = new FindOneAndUpdateOptions<BsonDocument>();`
`options.ReturnDocument = ReturnDocument.After;`. 
It doesn't return the document by default. Hence, before writing any implementation there should have been a dedicated test testing whether the gateway returns what you expected - it would fail due to default behaviour. Then you'd add these lines, and it would pass.
I also mentioned that there's a need for test that checks whether the document in the database has been updated. Since you're writing the test prior to implementation - you're treating the gateway as a black box. In order to check if the document in the db was updated - you need to check the document in the db. You can't trust the output of the gateway for that (which is a separate test in itself).
Think of this scenario: 
What if, someone decided that it's better to build up entire object within gateway first, then try to update it, and then return it. In the unlikely, but possible case - if he did the update step wrong the following happens:
1) He builds up a correct object
2) He tries to do the updating, but a failure happens due to mistake in the implementation (further, it's possible to mess up catch blocks and not get notified of this failure)
3) He returns the correctly built up object from step 1
Since the current gateway tests only checks for the output - in this case, it would never notice that the document was never updated within the database.
In short the current test never checks inside the actual database, the trustworthiness of the test relies on gateway being implemented correctly, which defeats the purpose of having the test.

However, since the implementation is already there, and given the scenario of test not catching up on failure is unlikely (even though possible) - this is not a high priority right now (since it doesn't add much benefit) considering we're out of time to do any changes anymore.

File: mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs (hunk 17 above)
Comment (Duslerke) [reply mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs:67]: There is no need to use TestCases, when we can use random Guid. It just increase the number of tests for no good reason and takes longer to run them.

Reason why you don't need test cases is: 
1) The triangulation is need to make sure that the test isn't passing due to hardcoded value.
2) When you use random Guid, the probability of the Hardcoded Guid matching the random one is
~10^(-38) . Which coincides with probability of Guid collision. I mean, it is Guid collision.

You could say, there's still a probability of it happening, but... in that case I'd say that if that's your worry, then you shouldn't be trusting your Guid to be your unique identifier in the first place. If you're worried so much about Guid collision making test inaccurate you should worry equal amount about guid making your application not work as intended (because the probabilities of both are equal...well test being inaccurate is less likely due to there being only 1 hardcoded value rather 100s of documents).

File: mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs [hunk 19]
```diff
@@ -83,6 +88,65 @@ public void check_gateway_returns_expected_response()
             Assert.AreEqual(response, result.ProcessData);
         }
 
+        #region Update Process Data
+        [Test]
+        public void update_process_data_verify_gateway_calls_database_with_parameters()
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs:93]: Test name doesn't make sense - the name suggests that it verifies whether the gateway calls the database...
, but it's the test about whether the usecase calls the gateway. It would be nice to also have it test for the values inside arguments, not just argument types.

So maybe when we do refactor, we should change this.

File: mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs [hunk 20]
```diff
@@ -83,6 +88,65 @@ public void check_gateway_returns_expected_response()
...
+        {
+            //arrange            
+            var request = new UpdateProcessDataRequest() { processDataToUpdate = new MatUpdateProcessData() };
+            //act
+            var result = processDataUseCase.ExecuteUpdate(request);
+            //assert
+            mockMatGateway.Verify(v => v.UpdateProcessData(It.IsAny<UpdateDefinition<BsonDocument>>(),It.IsAny<string>()), Times.Once);
+        }
+        [Test]
+        public void update_process_data_gateway_call_returns_updateprocsesdataresponse_object()
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs:103]: Test name confusing - it seems as if the test is 
`"gateway call returns 'updateprocsesdataresponse' object"`
, but from implementation I can see that it's among the lines of:
`"after doing the gateway call, usecase returns 'updateprocsesdataresponse' object"`
, however I'm not convinced the `calling the gateway` part is even necessary.

File: mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs [hunk 21]
```diff
@@ -83,6 +88,65 @@ public void check_gateway_returns_expected_response()
...
+            //act
+            mockMatGateway.Setup(x => x.UpdateProcessData(It.IsAny<UpdateDefinition<BsonDocument>>(),It.IsAny<string>())).Returns(new MatProcessData());
+            var result = processDataUseCase.ExecuteUpdate(request);
+            //assert
+            Assert.IsInstanceOf<MatProcessData>(result.UpdatedProcessData);
+            Assert.IsInstanceOf<UpdateProcessDataResponse>(result);
+        }
+
+        [Test]
+        public void update_process_data_returns_updateprocessdataresponse_object()
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs:118]: Test name does not describe the test. This test is clearly more about checking whether the data persists after passing through the usecase, rather than simply checking if the object of specified _(in the test name)_ type is returned.

File: mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs [hunk 22]
```diff
@@ -83,6 +88,65 @@ public void check_gateway_returns_expected_response()
...
+            Assert.IsInstanceOf<UpdateProcessDataRequest>(result.Request);
+            Assert.IsInstanceOf<MatProcessData>(result.UpdatedProcessData);
+            Assert.AreEqual(JsonConvert.SerializeObject(new MatProcessData()), JsonConvert.SerializeObject(result.UpdatedProcessData));
+            Assert.NotNull(result.GeneratedAt);
+            Assert.IsInstanceOf<DateTime>(result.GeneratedAt);
+        }
+        [TestCase("123-456-789-000")]
+        [TestCase("acb5-abc-abc-abc")]
+        [TestCase("00000000-0000-0000-0000-000000000000")]
+        public void update_process_data_verify_gateway_calls_database_with_parameters(string processRef)
```
Comment (Duslerke) [inline mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs:138]: Why was this overload of a the first test in this #region added?

image

File: mat-process-api/V1/Validators/UpdateProcessDocumentRequestValidator.cs (hunk 16 above)
Comment (Duslerke) [reply mat-process-api/V1/Validators/UpdateProcessDocumentRequestValidator.cs:47]: I take it that yes means, it's intended behaviour. Then it's all good, otherwise we would have needed more tests to cover the validator.

File: mat-process-api/V1/Boundary/UpdateProcessDataRequest.cs (hunk 10 above)
Comment (Duslerke) [reply mat-process-api/V1/Boundary/UpdateProcessDataRequest.cs:13]: We agreed that this wrap up discussion should be left for when refactoring happens, as it doesn't add any difficult complications right now.

File: mat-process-api/V1/Controllers/ProcessDataController.cs (hunk 14 above)
Comment (Duslerke) [reply mat-process-api/V1/Controllers/ProcessDataController.cs:71]: I still feel that the POST endpoint should have a name for an input parameter specified in Swagger documentation. I don't think it's okay to just remove it because it's name matches another endpoints parameter name. But we can leave this for refactoring.

File: mat-process-api/V1/Controllers/ProcessDataController.cs [hunk 23]
```diff
@@ -54,21 +58,41 @@ public IActionResult GetProcessData(string processRef)
...
+
+                try
+                {
+                    _logger.LogInformation($"Update ProcessData request for process ID {updateRequest.processRef}");
+                    var result = _processDataUsecase.ExecuteUpdate(updateRequest);
+                    return Ok(result);
+                }
+                catch (DocumentNotFound ex)
+                {
+                    return StatusCode(200, $"Document with reference {updateRequest.processRef} was not found in the database." +
```
Comment (Duslerke) [inline mat-process-api/V1/Controllers/ProcessDataController.cs:81]: The "Success error" question has been left for a discussion some other time. 
It was decided to leave this as is, as there's a lack of time discuss something as minor as this.

File: mat-process-api/V1/Controllers/ProcessDataController.cs [hunk 24]
```diff
@@ -54,21 +58,41 @@ public IActionResult GetProcessData(string processRef)
         /// If processData needs to be update, it's instead replaced with a full new one.
         /// Returning flat out 200 on successful update
         /// </summary>
-        /// <param name="processDataJSON"></param>
+        /// <param name="updateRequest"></param>
         /// <returns></returns>
         [HttpPatch]
+        [Route("{processRef}")]
```
Comment (Duslerke) [inline mat-process-api/V1/Controllers/ProcessDataController.cs:64]: Is this parameter supposed to be captured or used anywhere? (Asking because it's not used)

File: mat-process-api/V1/Factories/ProcessDataFactory.cs [hunk 25]
```diff
@@ -26,7 +28,7 @@ public static MatProcessData CreateProcessDataObject(BsonDocument bsonResult)
 
         public static MatProcessData CreateProcessDataObject(PostInitialProcessDocumentRequest requestObject) //Maps PostInitialDocumentRequest object to ProcessData domain object
         {
-            DateTime dateOfCreation = DateTime.UtcNow; // need this here because DateCreated and DateLastModified have to be equal
+            DateTime dateOfCreation = DateTime.Now; // need this here because DateCreated and DateLastModified have to be equal
```
Comment (Duslerke) [inline mat-process-api/V1/Factories/ProcessDataFactory.cs:31]: Is there a reason why this changed from ".Now" into ".UtcNow", and then changed back to ".Now"?

**Review by Duslerke — APPROVED** [review]
It's really good the old Transactions stuff was removed - didn't expect that to happen this PR. 👍 
Really like that update query related code was moved into its own category - will make finding code easier.
Extra tests added 👍 

It would have been good if the consistency had been reached of using the "faker" rather than unnecessary TestCases. Also would have been nice if all the missing tests had been added.
Some other things involving test naming and other small things, we can leave for now as we're pressed for time.
